admin/server/device_code.go

```diff
@@ -0,0 +1,241 @@
+package server
+
+import (
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/rilldata/rill/admin/database"
+	"github.com/rilldata/rill/admin/server/auth"
+	"github.com/rilldata/rill/cli/pkg/deviceauth"
+)
+
+const (
+	tokenExpirationSeconds = 60 * 10 // 10 minutes
+	deviceCodeGrantType    = "urn:ietf:params:oauth:grant-type:device_code"
+)
+
+// DeviceCodeResponse encapsulates the response for obtaining a device code.
+type DeviceCodeResponse struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationURI         string `json:"verification_uri"`
+	VerificationCompleteURI string `json:"verification_uri_complete"`
+	ExpiresIn               int    `json:"expires_in"`
+	PollingInterval         int    `json:"interval"`
+}
+
+// TokenRequest encapsulates the request for obtaining an access token.
+type TokenRequest struct {
+	GrantType  string `json:"grant_type"`
+	DeviceCode string `json:"device_code"`
+	ClientID   string `json:"client_id"`
+}
+
+// handleDeviceCodeRequest creates a 24 digit random device code and 8 digit user code and returns that
+// to the client. The device code is used to poll for an access token, and the user code is displayed
+// to the user and is used to authorize the device. The device code and user code are stored in the
+// server's device code store.
+func (s *Server) handleDeviceCodeRequest(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		internalServerError(w, fmt.Errorf("failed to read request body: %w", err))
+		return
+	}
+	bodyStr := string(body)
+	values, err := url.ParseQuery(bodyStr)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	clientID := values.Get("client_id")
+	if clientID == "" {
+		http.Error(w, "client_id is required", http.StatusBadRequest)
+		return
+	}
+	scopes := strings.Split(values.Get("scope"), " ")
+	if len(scopes) == 0 {
+		http.Error(w, "scope is required", http.StatusBadRequest)
+		return
+	}
+	if len(scopes) > 1 || scopes[0] != "full_account" {
+		http.Error(w, "invalid scope", http.StatusBadRequest)
+		return
+	}
+	authCode, err := s.admin.IssueAuthCode(r.Context(), clientID)
+	if err != nil {
+		internalServerError(w, fmt.Errorf("failed to issue auth code: %w", err))
+		return
+	}
+
+	verificationURI := s.opts.DeviceVerificationHost + "/oauth/device"
+	resp := DeviceCodeResponse{
+		DeviceCode:              authCode.DeviceCode,
+		UserCode:                authCode.UserCode,
+		VerificationURI:         verificationURI,
+		VerificationCompleteURI: verificationURI + "?user_code=" + authCode.UserCode,
+		ExpiresIn:               tokenExpirationSeconds,
+		PollingInterval:         5,
+	}
+	e := json.NewEncoder(w)
+	e.SetIndent("", "  ")
+	err = e.Encode(resp)
+	if err != nil {
```

Comment:
Just use normal json.Marshal – you can get a browser plugin to format JSON if you want